### PR TITLE
float to/from bits and classify: update for float semantics RFC

### PR DESCRIPTION
**Pull request by RalfJung**
With https://github.com/rust-lang/rfcs/pull/3514 having been accepted, it is clear that hardware which e.g. flushes subnormal to zero is just non-conformant from a Rust perspective -- this is a hardware bug, or maybe an LLVM backend bug (where LLVM doesn't lower floating-point ops in a way that they have the standardized behavior). So update the comments here to make it clear that we don't have to do any of this, we're just being nice.

Also remove the subnormal/NaN checks from the (unstable) const-version of to/from-bits; they are not needed since we decided with the aforementioned RFC that it is okay to get a different result at const-time and at run-time.

r? @workingjubilee since I think you wrote many of the comments I am editing here.

### CODE REVIEW

**Comment by RalfJung** [comment]
@rust-lang/wg-const-eval this also makes it so that to_bits / from_bits in const will expose the bit pattern of NaNs (and subnormals). That matches the accepted float semantics RFC: the exposed bit pattern is guaranteed to be *some* NaN bit pattern, but which exact bit pattern one gets is not fixed. (These functions are still unstable, though they could be stabilized soon.)

**Comment by tgross35** [comment]
Would you be able to update f16 and f128? If you run into anything too weird there then I can do it separately after this merges.

(` 5 files changed, 43 insertions(+), 394 deletions(-)` love these kind of diffstats)

**Comment by RalfJung** [comment]
I'll wait for a review before doing that so that I don't have to do the same edits 4 times.

**Review by workingjubilee — APPROVED** [review]
cool, okay!

sorry for taking so long to review but I had to like... also go over the float rfc again with a fine-tooth comb.

File: tests/ui/consts/const-float-bits-conv.rs
Comment (RalfJung) [reply tests/ui/consts/const-float-bits-conv.rs:44]: This is not new code, I just moved it. ;)

File: library/core/src/num/f16.rs
Comment (RalfJung) [inline library/core/src/num/f16.rs:454]: @tgross35 f16 is somewhat odd here -- it checks for infinity above and then again with the bitwise check here. That's different from f32, which only checks it separately, and f64, which only checks it via the bits. Isn't this arm here dead code because we already checked `is_infinite` above?

**Comment by RalfJung** [comment]
> ( 5 files changed, 43 insertions(+), 394 deletions(-) love these kind of diffstats)

Yeah same. :D
FWIW in principle we could delete a lot more code here and just always to bit-pattern classification. According to our own semantics, that is correct. The LLVM x86-non-SSE codegen path is just buggy (and unlikely to be fixed as nobody who knows how to fix it cares enough bout this target) and that's why we are carrying these work-arounds...

File: library/core/src/num/f16.rs
Comment (tgross35) [reply library/core/src/num/f16.rs:456]: You are right, that is redundant - I don't think that was done with any specific intent.

Comment (RalfJung) [reply library/core/src/num/f16.rs:456]: I will follow what f32 does then since it seems closest to f16.

File: library/core/src/num/f32.rs
Comment (workingjubilee) [inline library/core/src/num/f32.rs:680]: ugh, sorry, noticed this a bit late.
```suggestion
            // So, this does use bitpattern matching for the rest. On x87, due to the incorrect
```

File: library/core/src/num/f64.rs
Comment (workingjubilee) [inline library/core/src/num/f64.rs:677]:
```suggestion
            // So, this does use bitpattern matching for the rest. On x87, due to the incorrect
```

File: library/core/src/num/f16.rs
Comment (workingjubilee) [inline library/core/src/num/f16.rs:449]:
```suggestion
            // So, this does use bitpattern matching for the rest. On x87, due to the incorrect
```

**Review by workingjubilee — COMMENTED** [review]
sorry, only noticed this typo on this last scan.

**Comment by RalfJung** [comment]
Typo fixed :)

**Comment by workingjubilee** [comment]
huzzah!

looks good.
@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r-
failed in https://github.com/rust-lang/rust/pull/129158

**Comment by workingjubilee** [comment]
what lol

**Comment by workingjubilee** [comment]
```
2024-08-16T17:51:54.2269181Z ---- [ui] tests/ui/consts/const-float-bits-conv.rs stdout ----
2024-08-16T17:51:54.2269730Z 
2024-08-16T17:51:54.2278169Z error: test run failed!
2024-08-16T17:51:54.2278710Z status: exit status: 101
2024-08-16T17:51:54.2281486Z command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-float-bits-conv" && RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUST_TEST_THREADS="4" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-float-bits-conv/a"
2024-08-16T17:51:54.2283801Z stdout: none
2024-08-16T17:51:54.2284357Z --- stderr -------------------------------
2024-08-16T17:51:54.2285439Z thread 'main' panicked at /checkout/tests/ui/consts/const-float-bits-conv.rs:51:5:
2024-08-16T17:51:54.2286463Z assertion `left == right` failed
2024-08-16T17:51:54.2287071Z   left: 2144687445
2024-08-16T17:51:54.2287556Z  right: 2140493141
2024-08-16T17:51:54.2288336Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-08-16T17:51:54.2289406Z ------------------------------------------
2024-08-16T17:51:54.2289856Z 
2024-08-16T17:51:54.2289867Z 
2024-08-16T17:51:54.2290077Z 
2024-08-16T17:51:54.2290257Z failures:
2024-08-16T17:51:54.2290879Z     [ui] tests/ui/consts/const-float-bits-conv.rs
2024-08-16T17:51:54.2291396Z 
```
awesome.

**Comment by workingjubilee** [comment]
it looks like that's what's failing...?
```rust
    const_assert!(f32::from_bits(MASKED_NAN2).to_bits(), MASKED_NAN2);
```

**Comment by RalfJung** [comment]
Ah so maybe inlining the function made the check not work properly any more?

IMHO we should just disable the test on this target as that's a (known, documented) target bug. But @workingjubilee you spent a bunch of effort trying to make these things work on broken targets so if you prefer I'll un-inline that function again.

**Comment by tgross35** [comment]
The test could probably be fixed to use either `black_box` or `const { ... }`  to force the intended behavior

**Comment by RalfJung** [comment]
Yeah I just realized, this is just using `from_bits`/`to_bits`. All I did there is remove the const-eval special case. But it seems that simplified the code far enough that it's now inlined differently, or so?

**Comment by RalfJung** [comment]
Ah! Good point.

Okay, this should fix it then -- I just have it skip the problematic comparison on i586 targets.
@bors r=workingjubilee

**Review by beetrees — CHANGES_REQUESTED** [review]

**Comment by workingjubilee** [comment]
thank you ecstatic-morse for explicitly documenting this test is about preserving bits for simple casts even in the signaling/quieted cases, you just saved me from making a very bad suggestion.

File: tests/ui/consts/const-float-bits-conv.rs
Comment (workingjubilee) [reply tests/ui/consts/const-float-bits-conv.rs:86]: let's rename these to SIGNALING_NAN and QUIET_NAN shall we?

I am aware of the mips differentiation, you can comment on "according to IEEE754-**2008**" if you like.

File: tests/ui/consts/const-float-bits-conv.rs
Comment (workingjubilee) [inline tests/ui/consts/const-float-bits-conv.rs:79]:
```suggestion
    // NOTE: These names assume `f{BITS}::NAN` is a quiet NAN and IEEE754-2008's NaN rules apply!
    const QUIET_NAN: u64 = f64::NAN.to_bits() ^ 0x0005_5555_5555_5555;
    const SIGNALING_NAN: u64 = f64::NAN.to_bits() ^ 0x000A_AAAA_AAAA_AAAA;
```

File: tests/ui/consts/const-float-bits-conv.rs
Comment (workingjubilee) [inline tests/ui/consts/const-float-bits-conv.rs:50]:
```suggestion
    // NOTE: These names assume `f{BITS}::NAN` is a quiet NAN and IEEE754-2008's NaN rules apply!
    const QUIET_NAN: u32 = f32::NAN.to_bits() ^ 0x002A_AAAA;
    const SIGNALING_NAN: u32 = f32::NAN.to_bits() ^ 0x0055_5555;
```

File: tests/ui/consts/const-float-bits-conv.rs
Comment (workingjubilee) [inline tests/ui/consts/const-float-bits-conv.rs:85]:
```suggestion
    const_assert!(f64::from_bits(QUIET_NAN).is_nan());
    const_assert!(f64::from_bits(SIGNALING_NAN).is_nan());
    const_assert!(f64::from_bits(QUIET_NAN).to_bits(), QUIET_NAN);
    if !has_broken_floats() {
        const_assert!(f64::from_bits(SIGNALING_NAN).to_bits(), SIGNALING_NAN);
```

File: tests/ui/consts/const-float-bits-conv.rs
Comment (workingjubilee) [inline tests/ui/consts/const-float-bits-conv.rs:56]:
```suggestion
    const_assert!(f32::from_bits(QUIET_NAN).is_nan());
    const_assert!(f32::from_bits(SIGNALING_NAN).is_nan());
    const_assert!(f32::from_bits(QUIET_NAN).to_bits(), QUIET_NAN);
    if !has_broken_floats() {
        const_assert!(f32::from_bits(SIGNALING_NAN).to_bits(), SIGNALING_NAN);
```

**Review by workingjubilee — COMMENTED** [review]
Like this, or so

**Pull request merged**
